network/test/middleware_test.go

```diff
@@ -805,6 +812,212 @@ func (m *MiddlewareTestSuite) TestUnsubscribe() {
+		Return(nil).
+		Run(func(args mockery.Arguments) {
+			event := args[2].(*libp2pmessage.TestMessage)
+			mu.Lock()
+			receivedMsgs[event.Text] = true
+			mu.Unlock()
+		})
+
+	var protocol protocol.ID
+
+	streamCount := maxStreams * 2
+
+	// this test only works if streamCount > mw.unicastMaxStreamsPerPeer
+	require.Greater(m.T(), streamCount, maxStreams)
+
+	streams := make(chan streamEvent, streamCount)
+	for i := 0; i < streamCount; i++ {
+		s, err := m.nodes[sender].CreateStream(m.mwCtx, receiverPeerID)
+		require.NoError(m.T(), err)
+
+		e := streamEvent{
+			message: fmt.Sprintf("message %d", i),
+			stream:  s,
+		}
+
+		streams <- e
+
+		// write message to stream (don't use SendDirect because it will close the stream)
+		msg, _, _ := messageutils.CreateMessage(m.T(), senderNodeID, receiverNodeID, testChannel, e.message)
+		bufw := bufio.NewWriter(s)
+		writer := ggio.NewDelimitedWriter(bufw)
+
+		err = writer.WriteMsg(msg)
+		require.NoError(m.T(), err)
+
+		err = bufw.Flush()
+		require.NoError(m.T(), err)
+	}
+	close(streams)
+
+	// give the receiver time to process the messages
+	time.Sleep(50 * time.Millisecond)
+
+	// make sure there are no more than the max number of streams still open (the rest were reset)
+	count := p2putils.CountStreams(m.nodes[receiver].Host(), senderPeerID, protocol, libp2pnet.DirInbound)
+	assert.LessOrEqual(m.T(), count, maxStreams)
+
+	// close all streams. Streams that were reset will return an error
+	// stream processing may happen out of order, so just count up all the resets
+	resets := 0
+	closedStreams := 0
+	for e := range streams {
+		err := e.stream.Close()
+		if err != nil {
+			assert.True(m.T(), strings.Contains(err.Error(), "stream reset"))
+			assert.False(m.T(), receivedMsgs[e.message], "message '%s' should not have been received", e.message)
+			resets++
+		} else {
+			assert.True(m.T(), receivedMsgs[e.message], "message '%s' should have been received", e.message)
+		}
+		closedStreams++
+	}
+	assert.Equal(m.T(), streamCount, closedStreams)
+
+	// middleware should only accept up to mw.unicastMaxStreamsPerPeer messages. it's possible there
+	// are additional resets due to flakiness in the network setup, but make sure at least 90% of the
+	// expected streams were accepted
+	assert.LessOrEqual(m.T(), len(receivedMsgs), maxStreams)
+	assert.GreaterOrEqual(m.T(), len(receivedMsgs), maxStreams*9/10)
+
+	// the rest should result in resets
+	assert.GreaterOrEqual(m.T(), resets, streamCount-maxStreams)
+}
+
+// TestUnicastStreamLimit_Concurrent tests that the middleware does not accept more than the maximum number of
+// unicast streams and streams created beyond the limit are reset. It does this by creating all of the streams
+// concurrently to additionally validate that the stream limit is enforced correctly.
+func (m *MiddlewareTestSuite) TestUnicastStreamLimit_Concurrent() {
```

Comment:
It looks like TestUnicastStreamLimit_Concurrent and TestUnicastStreamLimit have identical structures in the test setup only difference being the concurrency. Can you refactor the overlapping logic such that it is shared by both tests.